### DISCUSSION TITLE
Click to open URL in Markdown file?

### DISCUSSION THREAD

**Discussion by jtherrmann** [orbitalquark/textadept#418] [Q&A] [answered]
When editing a Markdown file, URLs are underlined (as expected), which suggests that they are recognized by the lexer. Therefore it should be possible to write a keybinding so that the user can Shift+click on a URL to open it in their default web browser. I'm new to Textadept and Lua, and not sure how to approach this. Would anyone be able to share a code snippet that would detect if the user has clicked on a particular Lexer element or, if this is too difficult, detect if the cursor is within a particular element?

**Comment by snoopy** [comment]
A quick glance at the docs revealed only a double click event but it might be a start.

```lua
events.connect(events.DOUBLE_CLICK, function(modifiers)
  if not (modifiers & view.MOD_SHIFT) then return end
  if buffer:get_lexer(true) ~= 'markdown' then return end
  textadept.editing.select_enclosed('(', ')')
  link = buffer:get_sel_text()
  if _G.WIN32 then
    os.spawn('start ' .. link)
  else
    os.spawn('open ' .. link)
  end
end)
```

**Comment by orbitalquark** [comment, marked as answer]
If you want a text region to respond to a click, you have to use an [indicator][1]. Since lexer elements are not marked with indicators, you'd have to write a function to search for regions to mark, mark them, and then write an `events.INDICATOR_CLICK` handler to respond. The following example is from an old quick reference book. It's not exactly what you need, but it illustrates the process:

```
-- Define hyperlink indicator.
local INDIC_LINK = _SCINTILLA.next_indic_number()
events.connect(events.VIEW_NEW, function()
	view.indic_hover_style[INDIC_LINK] = view.INDIC_TEXTFORE
	view.indic_hover_fore[INDIC_LINK] = 0xFF0000 -- blue
end)
-- Search the buffer and mark hyperlinks.
function mark_hyperlinks()
	local text = buffer:get_text()
	buffer.indicator_current = INDIC_LINK
	buffer:indicator_clear_range(1, buffer.length)
	for s, e in text:gmatch('()https?://%S+()') do
		buffer:indicator_fill_range(s, e - s)
	end
end
-- Open clicked hyperlinks in a web browser.
local browser_cmd = 'firefox "%s"'
events.connect(events.INDICATOR_CLICK, function(pos)
	local indicators = buffer:indicator_all_on_for(pos)
	if indicators & INDIC_LINK > 0 then
		local s = buffer:indicator_start(INDIC_LINK, pos)
		local e = buffer:indicator_end(INDIC_LINK, pos)
		local url = buffer:text_range(s, e + 1)
		os.spawn(browser_cmd:format(url))
	end
end)
```

[1]: https://orbitalquark.github.io/textadept/api.html#mark-text-with-indicators

Alternatively, you can write a routine to check if the caret is under a link, extract that link, and open it in your browser:

```
local link = buffer:style_of_name(lexer.LINK)
if buffer.style_at[buffer.current_pos] == link then
	local s, e = buffer.current_pos, buffer.current_pos
	for i = buffer.current_pos - 1, 1, -1 do
		if buffer.style_at[i] ~= link then break end
		s = s - 1
	end
	for i = buffer.current_pos, buffer.length do
		if buffer.style_at[i] ~= link then break end
		e = e + 1
	end
	local link_text = buffer:text_range(s, e)
	-- os.spawn() your browser with link_text here
end
```

I haven't tested either of these code snippets, but that's the idea.

**Reply by jtherrmann** [reply]
Cool, thanks! I'm not sure when I'll have time to try this out, but thanks for getting me started!
